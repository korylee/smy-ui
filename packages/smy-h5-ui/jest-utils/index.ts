import Vue from 'vue'
import { Wrapper } from '@vue/test-utils'
import { IN_BROWSER } from '@smy-h5/shared'

Vue.config.productionTip = false

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0,
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })
  el.dispatchEvent(event)

  return Vue.nextTick()
}

export async function triggerDrag(el: HTMLElement, relativeX: number, relativeY: number) {
  let x = relativeX
  let y = relativeY
  let startX = 0
  let startY = 0
  if (relativeX < 0) {
    startX = Math.abs(relativeX)
    x = 0
  }
  if (relativeY < 0) {
    startY = Math.abs(relativeY)
    y = 0
  }
  await trigger(el, 'touchstart', startX, startY)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await delay(300)
  await trigger(el, 'touchend', x, y)
}

export function mockConsole(method: keyof Console, fn: any = () => {}) {
  const originMethod = console[method] as any
  console[method] = fn
  return () => {
    console[method] = originMethod
  }
}

export function mockGetBoundingClientRect(rect: Partial<DOMRect>): () => void {
  if (IN_BROWSER) {
    const spy = jest.spyOn(Element.prototype, 'getBoundingClientRect')
    spy.mockReturnValue(rect as DOMRect)
    return () => spy.mockRestore()
  }
  return () => {}
}
