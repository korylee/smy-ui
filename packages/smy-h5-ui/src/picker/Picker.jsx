import SmyPopup from '../popup'
import { props, pickerToolbarPropKeys, pickerPopupPropKeys } from './props'
import { toPxNum } from '../_utils/dom'
import { assign, createGetPropertyFromItem, pick, toNumber, wrapInArray } from '../_utils/shared'
import PickerColumn from './PickerColumn.vue'
import { isArray, isNil } from '../_utils/is'
import { bem, name, looseEqual, pickerToolbarSlots, pickerPopupListeners, findIndexFromColumn } from './utils'
import { getListeners } from '../_mixins/listeners'
import { createChildrenMixin } from '../_mixins/relation'
import PickerToolbar from './PickerToolbar.vue'

import '../popup/popup.less'
import './picker.less'

let sid = 0

export default {
  name,
  components: { PickerColumn, SmyPopup, PickerToolbar },
  mixins: [createChildrenMixin('pickerGroup')],
  props,

  data: () => ({ pickedIndexes: [], pickedValues: [], scrollColumns: [] }),

  computed: {
    localOptionHeight({ optionHeight }) {
      return toPxNum(optionHeight)
    },
    center({ columnHeight, localOptionHeight }) {
      return (columnHeight - localOptionHeight) / 2
    },
    columnHeight({ optionCount, localOptionHeight }) {
      return toNumber(optionCount) * localOptionHeight
    },
    getText: (vm) => createGetPropertyFromItem(vm.itemText),
    getValue: (vm) => createGetPropertyFromItem(vm.itemValue),
    getChildren: (vm) => createGetPropertyFromItem(vm.itemChildren),
    getDisabled: (vm) => createGetPropertyFromItem(vm.itemDisabled, false),
  },

  watch: {
    value(val, oldVal) {
      if (this.initing) return
      if (looseEqual(val, oldVal) || looseEqual(val, this.pickedValues)) {
        return
      }
      const { scrollColumns } = this
      val?.forEach((item, index) => {
        const changed = item !== oldVal?.[index]
        const scrollColumn = scrollColumns[index]
        changed && scrollColumn && this.updatePicked(scrollColumn)
      })
    },
    columns: {
      immediate: true,
      handler(val, oldVal) {
        if (looseEqual(val, oldVal)) return
        this.init()
      },
    },
  },

  methods: {
    confirm() {
      this.stopScroll()

      const { pickedIndexes, pickedValues } = this
      this.$emit('confirm', pickedValues.slice(), pickedIndexes)
      this.$emit('input', pickedValues.slice())
    },

    cancel() {
      this.stopScroll()
      const { pickedIndexes, pickedValues } = this

      this.$emit('cancel', pickedValues, pickedIndexes)
    },

    init() {
      const { columns } = this
      this.initing = true
      this.pickedIndexes = []
      this.pickedValues = []
      this.scrollColumns = this.cascade
        ? this.initCascade(columns)
        : this.initNormal(isArray(columns[0]) ? columns : [columns])
      this.initing = false
    },

    initNormal(columns) {
      return columns.map(this.createScrollColumn)
    },

    initCascade(columns) {
      const scrollColumns = []
      this.setChildren(scrollColumns, 0, columns)
      return scrollColumns
    },

    setChildren(scrollColumns, columnIndex, children) {
      if (!isArray(children) || !children.length) return
      const scrollColumn = this.createScrollColumn(children, columnIndex, children)
      const { pickedIndex } = scrollColumn
      this.$set(scrollColumns, columnIndex, scrollColumn)
      const nextColumnIndex = columnIndex + 1
      this.setChildren(scrollColumns, nextColumnIndex, this.getChildren(children[pickedIndex], scrollColumn))
    },

    updatePicked(scrollColumn) {
      const { value, pickedIndexes, pickedValues, presetValue } = this
      const { column, columnIndex, pickedIndex: oldPickedIndex } = scrollColumn
      const pickedValue = wrapInArray(value)[columnIndex]
      const values = column.map((item) => this.getValue(item, scrollColumn))
      let findedIndex = -1
      if (!isNil(pickedValue)) {
        findedIndex = values.indexOf(pickedValue)
      }
      if (findedIndex === -1) {
        if (presetValue) {
          const defaultValue = presetValue(scrollColumn)
          isNil(defaultValue) || (findedIndex = values.indexOf(defaultValue))
        } else if (!isNil(oldPickedIndex)) {
          findedIndex = oldPickedIndex
        }
      }
      const pickedIndex = findIndexFromColumn(findedIndex, scrollColumn, this.getDisabled)
      if (oldPickedIndex === pickedIndex) return
      pickedIndexes[columnIndex] = pickedIndex
      pickedValues[columnIndex] = values[pickedIndex]
      scrollColumn.pickedIndex = pickedIndex
    },

    createScrollColumn(column, columnIndex, columns) {
      let scrollColumn = column.$_scrollColumn
      if (!scrollColumn) {
        scrollColumn = {
          id: sid++,
          column,
          columnIndex,
          columns,
        }
        Object.defineProperty(column, '$_scrollColumn', {
          value: scrollColumn,
          enumerable: false,
          writable: false,
          configurable: false,
        })
      }
      this.updatePicked(scrollColumn)
      return scrollColumn
    },

    onScrollInfo(index) {
      const { scrollColumns, pickedIndexes, pickedValues } = this
      const scrollColumn = scrollColumns[index]
      const { pickedIndex, column } = scrollColumn
      pickedIndexes[index] = pickedIndex
      pickedValues[index] = this.getValue(column[pickedIndex], scrollColumn)
      this.$emit('update:value', pickedValues.slice())
      this.$emit('scroll-into', scrollColumn)
    },

    rebuildChild(scrollColumn) {
      const { scrollColumns } = this
      const { columns, columnIndex, pickedIndex } = scrollColumn
      const nextColumnIndex = columnIndex + 1
      const column = columns[pickedIndex]
      const children = this.getChildren(column, scrollColumn)
      this.setChildren(scrollColumns, nextColumnIndex, children)
    },

    stopScroll() {
      const { columnRefs } = this.$refs
      columnRefs.forEach((columnRef) => {
        if (!columnRef?.scrolling) return
        columnRef.stopScroll()
        const { scrollColumn } = columnRef
        this.change(scrollColumn)
      })
    },

    change(scrollColumn) {
      const {
        $refs: { columnRefs },
        cascade,
      } = this
      cascade && this.rebuildChild(scrollColumn)
      const hasScrolling = columnRefs.some((columnRef) => columnRef.scrolling)
      if (hasScrolling) return
      const { pickedIndexes, pickedValues } = this
      this.$emit('change', pickedValues.slice(), pickedIndexes)
    },
  },

  render() {
    /**@type {import('vue').default} */
    const vm = this
    const _h = vm.$createElement
    const c = vm._self._c || _h
    const {
      popup,
      pickerGroup,
      toolbar,
      columnHeight,
      scrollColumns,
      localOptionHeight,
      center,
      allowHtml,
      // method
      confirm,
      cancel,
      getText,
      change,
      onScrollInfo,
      getDisabled,
    } = vm

    const renderToolbar = () => {
      if (toolbar && !pickerGroup) {
        const fallback = () => {
          const fns = pickerToolbarSlots.map((key) => ({
            key: key,
            fn: () => [vm._t(key)],
            proxy: true,
          }))
          return [
            c(PickerToolbar, {
              on: { confirm, cancel },
              attrs: pick(vm.$props, pickerToolbarPropKeys),
              scopedSlots: vm._u(fns, null, true),
            }),
          ]
        }
        return vm._t('toolbar', fallback)
      }
      return vm._e()
    }

    const renderBody = () => {
      const domPropName = allowHtml ? 'innerHTML' : 'textContent'

      return (
        <div class={bem('columns')} style={{ height: `${columnHeight}px` }}>
          {scrollColumns.map((scrollColumn, index) => {
            const scopedSlots = vm._u([
              {
                key: 'item',
                fn({ item, index }) {
                  const itemFallback = () => [
                    c('div', { class: bem('text'), domProps: { [domPropName]: vm._s(getText(item, scrollColumn)) } }),
                  ]
                  return [vm._t('item', itemFallback, { item, index }, scrollColumn)]
                },
              },
            ])
            const attrs = {
              scrollColumn,
              center,
              height: localOptionHeight,
              getDisabled,
            }
            return c(PickerColumn, {
              ref: 'columnRefs',
              refInFor: true,
              key: scrollColumn.id,
              attrs,
              on: {
                change: () => change(scrollColumn),
                'scroll-into': () => onScrollInfo(index),
              },
              scopedSlots,
            })
          })}
          <div class={bem('picked')} style={{ top: `${center}px`, height: `${localOptionHeight}px` }} />
          <div class={bem('mask')} style={{ backgroundSize: `100% ${center}px` }} />
        </div>
      )
    }

    const children = [renderToolbar(), vm._t('top'), renderBody()]

    if (popup && !pickerGroup) {
      const attrs = assign(
        {
          contentClass: bem(),
          wrapperClass: bem('popup'),
          'smy-picker-cover': '',
          position: 'bottom',
        },
        pick(vm.$props, pickerPopupPropKeys)
      )
      return c(SmyPopup, { attrs, on: getListeners.call(vm, pickerPopupListeners) }, children, 2)
    }

    return c('div', { class: bem() }, children, 2)
  },
}
