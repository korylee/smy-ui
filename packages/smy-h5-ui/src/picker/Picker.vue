<template>
  <Popup
    :show.sync="internalShow"
    :teleport="teleport"
    :close-on-click-overlay="closeOnClickOverlay"
    smy-picker-cover
    position="bottom"
    class="smy-picker__popup"
    @click-overlay="$emit('click-overlay')"
    @open="$emit('open')"
    @opened="$emit('opened')"
    @close="$emit('close')"
    @closed="$emit('closed')"
    @route-change="$emit('route-change')"
  >
    <div class="smy-picker" v-bind="$attrs">
      <div class="smy-picker__toolbar">
        <slot name="cancel">
          <span class="smy-picker__cancel-button" smy-picker-cover @click="handleCancel">
            {{ cancelButtonText }}
          </span>
        </slot>
        <slot name="title">
          <div class="smy-picker__title">{{ title }}</div>
        </slot>
        <slot name="confirm">
          <span class="smy-picker__confirm-button" smy-picker-cover @click="handleConfirm">
            {{ confirmButtonText }}
          </span>
        </slot>
      </div>
      <slot name="top"></slot>
      <div class="smy-picker__columns" :style="{ height: `${columnHeight}px` }">
        <div
          class="smy-picker__column"
          v-for="scrollCol in scrollColumns"
          :key="scrollCol.id"
          @touchstart="handleTouchstart($event, scrollCol)"
          @touchmove.prevent="handleTouchmove($event, scrollCol)"
          @touchend="handleTouchend($event, scrollCol)"
        >
          <div
            class="smy-picker__scroller"
            ref="scrollEl"
            :style="getScrollerStyle(scrollCol)"
            @transitionend="handleTransitionend(scrollCol)"
          >
            <div
              v-for="(text, textIndex) in scrollCol.column.texts"
              :key="text"
              :class="{
                'smy-picker__option--picked': textIndex === scrollCol.index,
              }"
              :style="{
                height: `${localOptionHeight}px`,
              }"
              class="smy-picker__option"
            >
              <div class="smy-picker__text">
                {{ textFormatter(text, scrollCol.columnIndex) }}
              </div>
            </div>
          </div>
        </div>
        <div
          class="smy-picker__picked"
          :style="{
            top: `${center}px`,
            height: `${localOptionHeight}px`,
          }"
        ></div>
        <div
          class="smy-picker__mask"
          :style="{
            backgroundSize: `100% ${center}px`,
          }"
        ></div>
      </div>
    </div>
  </Popup>
</template>

<script>
import Popup from '../popup'
import { toPxNum } from '../_utils/shared'
import { props } from './props'

function getTranslate(el) {
  const { transform } = window.getComputedStyle(el)
  return +transform.slice(transform.lastIndexOf(',') + 2, transform.length - 1)
}

const MOMENTUM_RECORD_TIME = 300
const MOMENTUM_ALLOW_DISTANCE = 15

let sid = 0

export default {
  name: 'SmyPicker',
  components: { Popup },
  props,
  data: () => ({
    scrollColumns: [],
    prevIndexes: [],
  }),
  computed: {
    internalShow: {
      get() {
        return this.show
      },
      set(val) {
        this.$emit('update:show', val)
      },
    },
    localOptionHeight() {
      return toPxNum(this.optionHeight)
    },
    localOptionCount() {
      return toPxNum(this.optionCount)
    },
    center() {
      return (this.columnHeight - this.localOptionHeight) / 2
    },
    columnHeight() {
      return this.localOptionCount * this.localOptionHeight
    },
  },
  watch: {
    columns: {
      immediate: true,
      handler(val) {
        this.scrollColumns = this.normalizeNormalColumns(val)
        const { indexes } = this.getPicked()
        this.prevIndexes = [...indexes]
      },
    },
  },

  mounted() {
    this.setScrollEls()
    this.$on('hook:updated', this.setScrollEls)
  },

  methods: {
    handleConfirm() {
      this.stopScroll()

      const { texts, indexes } = this.getPicked()
      this.prevIndexes = [...indexes]
      this.$emit('confirm', texts, indexes)
    },
    initIndex() {
      this.scrollColumns.forEach((scrollCol) => {
        const index = scrollCol.column.initialIndex || 0
        scrollCol.index = index
        this.scrollTo(scrollCol, index, 200)
      })
    },
    handleCancel() {
      this.stopScroll()
      this.initIndex()

      const { texts, indexes } = this.getPicked()
      this.prevIndexes = [...indexes]
      this.$emit('cancel', texts, indexes)
    },
    handleTouchstart(e, scrollCol) {
      scrollCol.touching = true
      scrollCol.scrolling = false
      scrollCol.duration = 0
      scrollCol.translate = getTranslate(scrollCol.scrollEl)
    },
    handleTouchmove(event, scrollCol) {
      if (!scrollCol.touching) return

      const { clientY } = event.touches[0]
      const moveY = scrollCol.prevY !== undefined ? clientY - scrollCol.prevY : 0
      scrollCol.prevY = clientY
      scrollCol.translate += moveY

      this.limitTranslate(scrollCol)

      const now = performance.now()
      if (now - scrollCol.momentumTime > MOMENTUM_RECORD_TIME) {
        scrollCol.momentumTime = now
        scrollCol.momentumPrevY = scrollCol.translate
      }
    },
    handleTouchend(event, scrollCol) {
      scrollCol.touching = false
      scrollCol.scrolling = true
      scrollCol.prevY = undefined
      const distance = scrollCol.translate - scrollCol.momentumPrevY
      const duration = performance.now() - scrollCol.momentumTime
      const shouldMomentum = Math.abs(distance) >= MOMENTUM_ALLOW_DISTANCE && duration <= MOMENTUM_RECORD_TIME
      shouldMomentum && this.momentum(scrollCol, distance, duration)
      scrollCol.index = this.getIndex(scrollCol)
      this.scrollTo(scrollCol, scrollCol.index, shouldMomentum ? 1000 : 200)
    },
    handleTransitionend(scrollCol) {
      scrollCol.scrolling = false
      this.change(scrollCol)
    },
    stopScroll() {
      this.scrollColumns.forEach((scrollColumn) => {
        scrollColumn.translate = getTranslate(scrollColumn.scrollEl)
        scrollColumn.index = this.getIndex(scrollColumn)
        this.scrollTo(scrollColumn, scrollColumn.index, 0)
      })
    },
    getIndex(scrollCol) {
      const index = Math.round((this.center - scrollCol.translate) / this.localOptionHeight)
      return this.boundaryIndex(scrollCol, index)
    },
    setScrollEls() {
      const { scrollEl } = this.$refs
      if (!scrollEl) return

      this.scrollColumns.forEach((scrollColumn, index) => {
        scrollColumn.scrollEl = scrollEl[index]
      })
    },

    getScrollerStyle(scrollCol) {
      return {
        transform: `translate3d(0, ${scrollCol.translate}px, 0)`,
        transitionDuration: `${scrollCol.duration}ms`,
        transitionProperty: scrollCol.duration ? 'transform' : 'none',
      }
    },

    limitTranslate(scrollCol) {
      const START_LIMIT = this.localOptionHeight + this.center
      const END_LIMIT = this.center - scrollCol.column.texts.length * this.localOptionHeight

      if (scrollCol.translate >= START_LIMIT) {
        scrollCol.translate = START_LIMIT
      }
      if (scrollCol.translate <= END_LIMIT) {
        scrollCol.translate = END_LIMIT
      }
    },
    boundaryIndex(scrollColumn, index) {
      const { length } = scrollColumn.column.texts
      index = index >= length ? length - 1 : index
      index = index <= 0 ? 0 : index
      return index
    },
    momentum(scrollColumn, distance, duration) {
      scrollColumn.translate += (Math.abs(distance / duration) / 0.003) * (distance < 0 ? -1 : 1)
    },

    scrollTo(scrollColumn, index, duration, noEmit = false) {
      const translate = this.center - this.boundaryIndex(scrollColumn, index) * this.localOptionHeight

      if (translate === scrollColumn.translate) {
        scrollColumn.scrolling = false
        !noEmit && this.change(scrollColumn)
      }

      scrollColumn.translate = translate
      scrollColumn.duration = duration
    },
    normalizeNormalColumns(columns) {
      return columns.map((col, colIndex) => {
        const column = Array.isArray(col) ? { texts: col } : col
        const scollColumn = {
          id: sid++,
          prevY: undefined,
          momentumPrevY: undefined,
          momentumTime: 0,
          touching: false,
          translate: this.center,
          index: column.initialIndex || 0,
          column,
          columnIndex: colIndex,
          duration: 0,
          scrollEl: null,
          scrolling: false,
        }
        this.scrollTo(scollColumn, scollColumn.index, 200)
        return scollColumn
      })
    },
    getPicked() {
      const indexes = []
      const texts = this.scrollColumns.map((scrollCol) => {
        const { index } = scrollCol
        indexes.push(index)
        return scrollCol.column.texts[index]
      })
      // const indexes = this.scrollColumns.map((scrollColumn) => scrollColumn.index)
      return {
        texts,
        indexes,
      }
    },
    change(scrollColumn) {
      const hasScrolling = this.scrollColumns.some((scrollColumn) => scrollColumn.scrolling)
      if (hasScrolling) return
      const { texts, indexes } = this.getPicked()
      const samePicked = indexes.every((index, idx) => index === this.prevIndexes[idx])
      if (samePicked) return
      this.prevIndexes = [...indexes]
      this.$emit('change', texts, indexes)
    },
  },
}
</script>

<style lang="less">
@import './picker.less';
</style>
