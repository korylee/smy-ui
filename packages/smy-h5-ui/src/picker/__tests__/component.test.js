import { mount } from '@vue/test-utils'
import Vue from 'vue'
import Picker from '..'
import { mockConsole, triggerDrag } from '../../../jest-utils'
import example from '../example'

test('test picker compomemt plugin', () => {
  Vue.use(Picker.Component)
  expect(Vue.component(Picker.Component.name)).toBeTruthy()
})

test('test picker example', () => {
  const restore = mockConsole('log')
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
  restore()
})

const columns = [['A', 'B', 'C']]

test('test scroll up & onConfirm', async () => {
  const onConfirm = jest.fn()

  const wrapper = mount(Picker.Component, {
    propsData: { show: true, columns },
    listeners: { confirm: onConfirm },
  })
  expect(wrapper.html()).toMatchSnapshot()
  const { element } = wrapper.find('.smy-picker__column')

  await triggerDrag(element, 0, -600)
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.smy-picker__confirm-button').trigger('click')
  expect(onConfirm).lastCalledWith(['C'], [2])

  wrapper.destroy()
})
