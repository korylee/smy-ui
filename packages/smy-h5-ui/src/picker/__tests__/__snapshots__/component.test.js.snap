// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test picker example 1`] = `
"<div>
  <h4>函数调用</h4>
  <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">单列选择</div>
      <!---->
    </div>
    <!---->
  </div>
  <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">多列选择</div>
      <!---->
    </div>
    <!---->
  </div>
  <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">级联选择</div>
      <!---->
    </div>
    <!---->
  </div>
  <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">默认选中项</div>
      <!---->
    </div>
    <!---->
  </div>
  <h4>组件调用</h4>
  <div class=\\"app-demo-title\\">单列选择</div>
  <div>
    <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
      <!---->
      <div class=\\"smy-cell__content\\">
        <div class=\\"smy-cell__title\\">请选择</div>
        <!---->
      </div>
      <div class=\\"smy-cell__extra\\">popup: <div tabindex=\\"0\\" aria-checked=\\"true\\" role=\\"switch\\" class=\\"smy-switch smy-switch--active\\">
          <div class=\\"smy-switch__thumb\\">
            <!---->
          </div>
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <transition-stub name=\\"smy-fade\\">
      <div role=\\"dialog\\" tabindex=\\"0\\" class=\\"smy-popup smy--box smy-picker__popup\\" style=\\"z-index: 998; display: none;\\">
        <div class=\\"smy-popup__overlay\\" style=\\"z-index: 999;\\"></div>
        <transition-stub name=\\"smy-slide-bottom\\">
          <!---->
        </transition-stub>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-demo-title\\">多列选择</div>
  <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">请选择</div>
      <!---->
    </div>
    <div class=\\"smy-cell__extra\\">popup: <div tabindex=\\"0\\" aria-checked=\\"true\\" role=\\"switch\\" class=\\"smy-switch smy-switch--active\\">
        <div class=\\"smy-switch__thumb\\">
          <!---->
        </div>
        <!---->
        <!---->
      </div>
    </div>
  </div>
  <transition-stub name=\\"smy-fade\\">
    <div role=\\"dialog\\" tabindex=\\"0\\" class=\\"smy-popup smy--box smy-picker__popup\\" style=\\"z-index: 998; display: none;\\">
      <div class=\\"smy-popup__overlay\\" style=\\"z-index: 999;\\"></div>
      <transition-stub name=\\"smy-slide-bottom\\">
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
  <div class=\\"app-demo-title\\">级联选择</div>
  <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">请选择</div>
      <!---->
    </div>
    <div class=\\"smy-cell__extra\\">popup: <div tabindex=\\"0\\" aria-checked=\\"true\\" role=\\"switch\\" class=\\"smy-switch smy-switch--active\\">
        <div class=\\"smy-switch__thumb\\">
          <!---->
        </div>
        <!---->
        <!---->
      </div>
    </div>
  </div>
  <transition-stub name=\\"smy-fade\\">
    <div role=\\"dialog\\" tabindex=\\"0\\" class=\\"smy-popup smy--box smy-picker__popup\\" style=\\"z-index: 998; display: none;\\">
      <div class=\\"smy-popup__overlay\\" style=\\"z-index: 999;\\"></div>
      <transition-stub name=\\"smy-slide-bottom\\">
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
  <div class=\\"app-demo-title\\">默认选中项</div>
  <div class=\\"smy-cell smy-cell--borderless smy-cell--insert\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">Tuesday,Afternoon</div>
      <!---->
    </div>
    <div class=\\"smy-cell__extra\\">popup: <div tabindex=\\"0\\" aria-checked=\\"true\\" role=\\"switch\\" class=\\"smy-switch smy-switch--active\\">
        <div class=\\"smy-switch__thumb\\">
          <!---->
        </div>
        <!---->
        <!---->
      </div>
    </div>
  </div>
  <transition-stub name=\\"smy-fade\\">
    <div role=\\"dialog\\" tabindex=\\"0\\" class=\\"smy-popup smy--box smy-picker__popup\\" style=\\"z-index: 998; display: none;\\">
      <div class=\\"smy-popup__overlay\\" style=\\"z-index: 999;\\"></div>
      <transition-stub name=\\"smy-slide-bottom\\">
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test scroll up & onConfirm 1`] = `
"<transition-stub name=\\"smy-fade\\">
  <div role=\\"dialog\\" tabindex=\\"0\\" class=\\"smy-popup smy--box smy-picker__popup\\" style=\\"z-index: 1001;\\">
    <div class=\\"smy-popup__overlay\\" style=\\"z-index: 1002;\\"></div>
    <transition-stub name=\\"smy-slide-bottom\\">
      <div smy-picker-cover=\\"\\" class=\\"smy-popup__content smy-popup__content--bottom smy-picker\\" style=\\"z-index: 1003;\\">
        <div class=\\"smy-picker__toolbar\\"><span smy-picker-cover=\\"\\" class=\\"smy-picker__cancel\\">取消</span>
          <div class=\\"smy-picker__title\\"></div> <span smy-picker-cover=\\"\\" class=\\"smy-picker__confirm\\">确认</span></div>
        <div class=\\"smy-picker__columns\\" style=\\"height: 264px;\\">
          <div class=\\"smy-picker__column\\">
            <div class=\\"smy-picker__scroller\\" style=\\"transform: translate3d(0, 110px, 0); transition-duration: 0ms; transition-property: none;\\">
              <div class=\\"smy-picker__item\\" style=\\"height: 44px;\\">
                <div class=\\"smy-picker__text\\">A</div>
              </div>
              <div class=\\"smy-picker__item\\" style=\\"height: 44px;\\">
                <div class=\\"smy-picker__text\\">B</div>
              </div>
              <div class=\\"smy-picker__item\\" style=\\"height: 44px;\\">
                <div class=\\"smy-picker__text\\">C</div>
              </div>
            </div>
          </div>
          <div class=\\"smy-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
          <div class=\\"smy-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
        </div>
      </div>
    </transition-stub>
  </div>
</transition-stub>"
`;

exports[`test scroll up & onConfirm 2`] = `
"<transition-stub name=\\"smy-fade\\">
  <div role=\\"dialog\\" tabindex=\\"0\\" class=\\"smy-popup smy--box smy-picker__popup\\" style=\\"z-index: 1001;\\">
    <div class=\\"smy-popup__overlay\\" style=\\"z-index: 1002;\\"></div>
    <transition-stub name=\\"smy-slide-bottom\\">
      <div smy-picker-cover=\\"\\" class=\\"smy-popup__content smy-popup__content--bottom smy-picker\\" style=\\"z-index: 1003;\\">
        <div class=\\"smy-picker__toolbar\\"><span smy-picker-cover=\\"\\" class=\\"smy-picker__cancel\\">取消</span>
          <div class=\\"smy-picker__title\\"></div> <span smy-picker-cover=\\"\\" class=\\"smy-picker__confirm\\">确认</span></div>
        <div class=\\"smy-picker__columns\\" style=\\"height: 264px;\\">
          <div class=\\"smy-picker__column\\">
            <div class=\\"smy-picker__scroller\\" style=\\"transform: translate3d(0, 22px, 0); transition-duration: 200ms; transition-property: transform;\\">
              <div class=\\"smy-picker__item\\" style=\\"height: 44px;\\">
                <div class=\\"smy-picker__text\\">A</div>
              </div>
              <div class=\\"smy-picker__item\\" style=\\"height: 44px;\\">
                <div class=\\"smy-picker__text\\">B</div>
              </div>
              <div class=\\"smy-picker__item\\" style=\\"height: 44px;\\">
                <div class=\\"smy-picker__text\\">C</div>
              </div>
            </div>
          </div>
          <div class=\\"smy-picker__picked\\" style=\\"top: 110px; height: 44px;\\"></div>
          <div class=\\"smy-picker__mask\\" style=\\"background-size: 100% 110px;\\"></div>
        </div>
      </div>
    </transition-stub>
  </div>
</transition-stub>"
`;
